File: frontend/src/types/schedule.ts
export interface AssignedBill {
  bill_id: number;
  name: string;
  due_date: string;
  amount: string;
  status: "on_time" | "late_flagged" | "paid" | "skipped";
  instance_id: number | null;
  actual_amount: string | null;
  is_variable: boolean;
  category: string;
  placement_source: "inferred" | "manual" | "paid";
  manual_pay_date: string | null;
  sinking_fund_applied: string;
  sinking_fund_shortfall: string;
}

export interface SinkingFundContribution {
  bill_id: number;
  name: string;
  next_due_date: string;
  target_amount: string;
  saved_amount: string;
  contribution_amount: string;
  shortfall_amount: string;
  surplus_amount: string;
}

export interface PayPeriod {
  period_index: number;
  pay_date: string;
  original_pay_date: string;
  is_overridden: boolean;
  period_start: string;
  period_end: string;
  opening_balance: string;
  total_bills: string;
  total_sinking_funds: string;
  remaining_balance: string;
  flagged_bill_count: number;
  assigned_bills: AssignedBill[];
  sinking_fund_contributions: SinkingFundContribution[];
}

export interface ScheduleSummary {
  from_date: string;
  to_date: string;
  period_count: number;
  total_flagged_bills: number;
}

export interface ScheduleResponse {
  periods: PayPeriod[];
  summary: ScheduleSummary;
}

export interface RebalanceMove {
  bill_id: number;
  name: string;
  due_date: string;
  amount: string;
  from_pay_date: string;
  to_pay_date: string;
  from_period_remaining_before: string;
  from_period_remaining_after: string;
  source_remaining_before: string;
  source_remaining_after: string;
  reason: string;
}

export interface RebalancePreview {
  source_pay_date: string;
  source_remaining_before: string;
  source_remaining_after: string;
  moves: RebalanceMove[];
}

File: backend/app/schemas/schedule.py
from __future__ import annotations

from datetime import date
from decimal import Decimal

from pydantic import BaseModel, Field


class AssignedBillOut(BaseModel):
    bill_id: int
    name: str
    due_date: date
    amount: Decimal
    status: str
    instance_id: int | None = None
    actual_amount: Decimal | None = None
    is_variable: bool = False
    category: str = "other"
    placement_source: str = "inferred"
    manual_pay_date: date | None = None
    is_carried_over: bool = False
    sinking_fund_applied: Decimal = Decimal("0")
    sinking_fund_shortfall: Decimal = Decimal("0")


class SinkingFundContributionOut(BaseModel):
    bill_id: int
    name: str
    next_due_date: date
    target_amount: Decimal
    saved_amount: Decimal
    contribution_amount: Decimal
    shortfall_amount: Decimal
    surplus_amount: Decimal


class PayPeriodOut(BaseModel):
    period_index: int
    pay_date: date
    original_pay_date: date
    is_overridden: bool = False
    period_start: date
    period_end: date
    opening_balance: Decimal
    total_bills: Decimal
    total_paid: Decimal = Decimal("0")
    total_unpaid: Decimal = Decimal("0")
    total_sinking_funds: Decimal = Decimal("0")
    remaining_balance: Decimal
    period_result: Decimal
    flagged_bill_count: int
    assigned_bills: list[AssignedBillOut]
    sinking_fund_contributions: list[SinkingFundContributionOut] = Field(
        default_factory=list
    )


class ScheduleSummary(BaseModel):
    from_date: date
    to_date: date
    period_count: int
    total_flagged_bills: int
    buffer_balance: Decimal = Decimal("0")


class ScheduleResponse(BaseModel):
    periods: list[PayPeriodOut]
    summary: ScheduleSummary


class RebalancePreviewRequest(BaseModel):
    source_pay_date: date
    max_moves: int = Field(default=5, ge=1, le=20)


class RebalanceMove(BaseModel):
    bill_id: int
    name: str
    due_date: date
    amount: Decimal
    from_pay_date: date
    to_pay_date: date
    from_period_remaining_before: Decimal
    from_period_remaining_after: Decimal
    source_remaining_before: Decimal
    source_remaining_after: Decimal
    reason: str


class RebalancePreviewResponse(BaseModel):
    source_pay_date: date
    source_remaining_before: Decimal
    source_remaining_after: Decimal
    moves: list[RebalanceMove]


class RebalanceApplyRequest(BaseModel):
    moves: list[RebalanceMove]


class SmoothingPreviewRequest(BaseModel):
    source_pay_date: date
